src/referral_maintenance.rs

```diff
@@ -27,20 +41,159 @@ pub async fn maintenaince_tasks(db: Arc<ReferralStore>, dune_data_folder: String
+            Ok(guard) => guard,
+            Err(poisoned) => poisoned.into_inner(),
+        };
+        uninitialized_app_data_hashes = guard
+            .app_data
+            .clone()
+            .into_iter()
+            .filter(|(_, referral)| {
+                // Unforunately, *referral != Referral::TryToFetchXTimes(_) exists only in nightly
+                *referral == Referral::TryToFetchXTimes(1)
+                    || *referral == Referral::TryToFetchXTimes(2)
+                    || *referral == Referral::TryToFetchXTimes(3)
+            })
+            .map(|(hash, _)| hash)
+            .collect();
+    }
+    // 3. try to retrieve all ipfs data for hashes and store them
+    for hash in uninitialized_app_data_hashes.iter() {
+        download_referral_from_ipfs_and_store_in_referral_store(db.clone(), *hash).await?;
+    }
     Ok(())
 }
 
-pub async fn referral_maintainance(memory_database: Arc<ReferralStore>, dune_data_folder: String) {
-    loop {
-        match maintenaince_tasks(Arc::clone(&memory_database), dune_data_folder.clone()).await {
-            Ok(_) => {}
-            Err(err) => tracing::debug!("Error during maintenaince_task for referral: {:?}", err),
+async fn download_referral_from_ipfs_and_store_in_referral_store(
+    db: Arc<ReferralStore>,
+    hash: H256,
+) -> Result<()> {
+    match get_cid_from_app_data(hash) {
+        Ok(cid) => {
+            tracing::debug!("cid for hash {:?} is {:?}", hash, cid);
+            match get_ipfs_file_and_read_referrer(cid.clone()).await {
+                Ok(referrer) => {
+                    tracing::debug!("Adding the referrer {:?} for the hash {:?}", referrer, hash);
+                    let mut guard = match db.0.lock() {
+                        Ok(guard) => guard,
+                        Err(poisoned) => poisoned.into_inner(),
+                    };
+                    guard.app_data.insert(hash, Referral::Address(referrer));
+                }
+                Err(err) => {
+                    tracing::debug!(
+                        "Could not find referrer in cid {:?}, due to the error {:?}",
+                        cid,
+                        err
+                    );
+                    let mut guard = match db.0.lock() {
+                        Ok(guard) => guard,
+                        Err(poisoned) => poisoned.into_inner(),
+                    };
+                    guard.app_data.entry(hash).and_modify(|referral_entry| {
+                        *referral_entry = match referral_entry.clone() {
+                            Referral::TryToFetchXTimes(x) => {
+                                if x > 1u64 {
+                                    Referral::TryToFetchXTimes(x - 1)
+                                } else {
+                                    Referral::Address(None)
+                                }
+                            }
+                            _ => Referral::Address(None),
```

Comment:
This kind of retry logic seems like it could be better suited outside the struct itself. Is there something special I am not understanding about this?